src/mpi/coll/op/opequal.c

```diff
@@ -0,0 +1,241 @@
+/* Values of is_equal: 1 equal 0 not equal, init using any value other than 0.*/
+#define MPIR_EQUAL_C_CASE_INT(c_type_) {                \
+        c_type_ *a = (c_type_ *)inoutvec;               \
+        c_type_ *b = (c_type_ *)invec;                  \
+        for (i = 0; i < len; ++i) {                         \
+            if (0 == b[i].is_equal || 0 == a[i].is_equal){    \
+                a[i].is_equal = 0;            			\
+            }else if (a[i].value != b[i].value){        \
+                a[i].is_equal = 0;                       \
+            }else{                                      \
+                a[i].is_equal = 1;                       \
+            }                                           \
+        }                                               \
+    }                                                   \
+    break
+
+#define MPIR_EQUAL_C_CASE_FLOAT(c_type_) {              \
+        c_type_ *a = (c_type_ *)inoutvec;               \
+        c_type_ *b = (c_type_ *)invec;                  \
+        for (i = 0; i < len; ++i) {                         \
+            if (0 == b[i].is_equal || 0 == a[i].is_equal){    \
+                a[i].is_equal = 0;            			\
+            }else if (!MPIR_EQUAL_FLOAT_COMPARE(a[i].value, b[i].value)){        \
+                a[i].is_equal = 0;                       \
+            }else{                                      \
+                a[i].is_equal = 1;                       \
+            }                                           \
+        }                                               \
+    }                                                   \
+    break
+
+#define MPIR_EQUAL_F_CASE(f_type_) {                    \
+        f_type_ *a = (f_type_ *)inoutvec;               \
+        f_type_ *b = (f_type_ *)invec;                  \
+        for (i = 0; i < flen; i += 2) {                       \
+            if(0 == b[i+1] || 0 == a[i+1]){             \
+                a[i+1] = 0;                        		\
+            }else if (a[i] != b[i]){                    \
+                a[i+1] = 0;                             \
+            }else{                                      \
+                a[i+1] = 1;                             \
+            }                                           \
+        }                                               \
+    }                                                   \
+    break
+
+void MPIR_EQUAL_user_defined_datatype_compare(void *invec, void *inoutvec, int *Len, MPI_Datatype * type)
+{
+	int mpi_errno = MPI_SUCCESS;
+	int i, j, size, len = *Len, is_equal; 
+	int data_len = 0, type_len = 0;
+	int num_ints, num_adds, num_types, combiner, *ints;
+	void *invec_i_pos, *invec_i_bool_pos, *inoutvec_i_pos, *inoutvec_i_bool_pos;
+	MPI_Aint *adds = NULL;
+	MPI_Aint lb, extent;
+	MPI_Datatype *types;
+
+	 /* decode */
+	MPIR_Type_get_envelope(*type, &num_ints, &num_adds, &num_types, &combiner);
```

Comment:
I thought this is a `static` function, how did you compile it? Use `MPIR_Type_get_envelope_impl` in stead.